lib/vrl/stdlib/src/parse_query_string.rs

```diff
@@ -0,0 +1,131 @@
+use bytes::Buf;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use url::form_urlencoded;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseQueryString;
+
+impl Function for ParseQueryString {
+    fn identifier(&self) -> &'static str {
+        "parse_query_string"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "parse query string",
+            source: r#"parse_query_string("foo=1&bar=2")"#,
+            result: Ok(r#"
+                {
+                    "foo": "1",
+                    "bar": "2"
+                }
+            "#),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        Ok(Box::new(ParseQueryStringFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::BYTES,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseQueryStringFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ParseQueryStringFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let query_string = self.value.resolve(ctx)?.try_bytes()?;
+        let mut result: BTreeMap<String, Value> = BTreeMap::new();
```

Comment:
```suggestion
        let mut result = BTreeMap::new();
```

The compiler _should_ be able to derive that this type is `BTreeMap<String, Value>`, so this type definition can be removed.

Reply:
It was complaining here:
```
|         let mut result = BTreeMap::new();
   |             ---------- consider giving `result` the explicit type `BTreeMap<std::string::String, V>`, where the type parameter `V` is specified
...
62 |                     if e.get().is_array() {
   |                          ^^^ cannot infer type for type parameter `V`
```

Reply:
Works now after refactoring.